### PR TITLE
<TreeView /> - Stage 3 - Keyboard Navigation

### PR DESCRIPTION
**Pull request by liorzisman**

### CODE REVIEW

File: test/components/tree-view.spec.tsx [hunk 1]
```diff
@@ -123,18 +122,223 @@ describe('<TreeView />', () => {
             const onSelectItem = sinon.spy();
             const { select } = clientRenderer.render(<TreeView dataSource={treeData} onSelectItem={onSelectItem}/>);
 
-            const nodeLabelToClick = getTreeItem(treeData[0].label);
+            selectItemWithLabel(select, treeData[0].label);
+
+            return waitFor(() => expect(onSelectItem).to.have.been.calledWithMatch(treeData[0]));
+        });
+
+        describe('Keyboard Navigation', () => {
```
Comment (JoMarton) [inline test/components/tree-view.spec.tsx]: it seems like there's A LOT of shared code between all the keyboard tests... should extract it to a method where possible

File: test/components/tree-view.spec.tsx [hunk 2]
```diff
@@ -123,18 +122,223 @@ describe('<TreeView />', () => {
...
 
-            return waitFor(() => {
-                expect(onSelectItem).to.have.been.calledOnce;
-                expect(onSelectItem).to.have.been.calledWithMatch(treeData[0]);
+                simulate.keyDown(select('TREE_VIEW_DEMO', 'TREE_VIEW'), { keyCode: KeyCodes.LEFT });
+
+                return waitForDom(() => expect(select(getTreeItem(treeData[0].label))).to.have.attr('data-focused', 'true'));
+            });
+
+            it('moves to child to if there is one after expanding the element if possible when right is clicked', async () => {
```
Comment (JoMarton) [inline test/components/tree-view.spec.tsx]: honestly I don't understand what this sentence means...

File: test/components/tree-view.spec.tsx [hunk 3]
```diff
@@ -123,18 +122,223 @@ describe('<TreeView />', () => {
...
+                await waitForDom(() => expect(select(getTreeItem(nodeChildren![0].label))).to.be.present());
+
+                await waitForDom(() => expect(select(getTreeItem(treeData[0].label))).to.have.attr('data-focused', 'true'));
+
+                simulate.keyDown(select('TREE_VIEW_DEMO', 'TREE_VIEW'), { keyCode: KeyCodes.RIGHT });
+
+                return waitForDom(() => expect(select(getTreeItem(nodeChildren![0].label))).to.have.attr('data-focused', 'true'));
+            });
+
+            it('focuses next are previous when down and up arrows are clicked', async () => {
```
Comment (JoMarton) [inline test/components/tree-view.spec.tsx]: again, don't think this it sentence makes sense

File: src/components/tree-view/tree-view.tsx [hunk 4]
```diff
@@ -57,58 +63,198 @@ export const TreeItem: React.SFC<TreeItemProps> = SBStateless(({ item, itemRende
...
+    focusNext = (item: TreeItemData) => this.onFocusItem!(this.getNextItem(item) as TreeItemData);
+    focusFirst = () => this.props.onFocusItem!(this.props.dataSource[0]);
+    focusLast = () =>
+        this.props.onFocusItem!(this.getLastAvailableItem(this.props.dataSource[this.props.dataSource.length - 1] as TreeItemData));
+
+    @action
+    onKeyDown = (e: any) => {
+        if (!this.props.focusedItem) return;
+
+        switch(e.keyCode) {
```
Comment (JoMarton) [inline src/components/tree-view/tree-view.tsx]: in switch we use `break`, not `return`

File: src/components/tree-view/tree-view.tsx [hunk 5]
```diff
@@ -57,58 +63,198 @@ export const TreeItem: React.SFC<TreeItemProps> = SBStateless(({ item, itemRende
...
+            this.parentsMap.set(item, parent);
+            this.initParentsMap(item.children || [], item);
         });
     }
 
     componentDidMount() {
         autorun(() => {
             if (this.props.selectedItem) {
-                this.stateMap.get(this.props.selectedItem)!.isSelected = true;
+                action(() => this.stateMap.getItemState(this.props.selectedItem!).isSelected = true)();
```
Comment (JoMarton) [inline src/components/tree-view/tree-view.tsx]: I think having methods for getting current selected / focused items state will make the code more readable

File: src/components/tree-view/tree-view.tsx [hunk 6]
```diff
@@ -57,58 +63,198 @@ export const TreeItem: React.SFC<TreeItemProps> = SBStateless(({ item, itemRende
...
+            default:
+                return;
+        }
     };
 
     render() {
         const TreeNode = this.props.itemRenderer!;
         return (
-            <div data-automation-id='TREE_VIEW' className="tree-view">
+            <div data-automation-id='TREE_VIEW' className="tree-view" tabIndex={0} onKeyDown={this.onKeyDown}>
```
Comment (JoMarton) [inline src/components/tree-view/tree-view.tsx]: please use `root` function from wix-react-tools. see example here:
https://github.com/wix/stylable-components/blob/master/src/components/checkbox/checkbox.tsx#L86

File: src/components/tree-view/tree-view.tsx (hunk 4 above)
Comment (liorzisman) [inline src/components/tree-view/tree-view.tsx]: done

File: src/components/tree-view/tree-view.tsx (hunk 6 above)
Comment (liorzisman) [inline src/components/tree-view/tree-view.tsx]: will do

File: test/components/tree-view.spec.tsx [hunk 7]
```diff
@@ -123,18 +123,223 @@ describe('<TreeView />', () => {
...
-            simulate.click(select(nodeLabelToClick));
+                simulate.keyDown(select('TREE_VIEW_DEMO', 'TREE_VIEW'), { keyCode: KeyCodes.LEFT });
 
-            return waitFor(() => {
-                expect(onSelectItem).to.have.been.calledOnce;
-                expect(onSelectItem).to.have.been.calledWithMatch(treeData[0]);
+                return waitForDom(() => expect(select(getTreeItem(treeData[0].label))).to.have.attr('data-focused', 'true'));
+            });
+
+            it('moves to child to if there is one after expanding the element if possible when right is clicked', async () => {
```
Comment (alisey) [inline test/components/tree-view.spec.tsx]: How about: "when the keyboard focus is on an expanded parent, pressing the right arrow key moves the focus to its first child"?

File: test/components/tree-view.spec.tsx (hunk 1 above)
Comment (liorzisman) [inline test/components/tree-view.spec.tsx]: modifications have been made

File: test/components/tree-view.spec.tsx (hunk 3 above)
Comment (liorzisman) [inline test/components/tree-view.spec.tsx]: fixed typo

File: test/utils/mobx.config.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+import { useStrict } from 'mobx';
+
+useStrict(false);
```
Comment (liorzisman) [inline test/utils/mobx.config.ts]: in order not to break everyone's code, after discussion, this should be done in 'horizontal' manner

**Comment by tyv** [comment]
this PR has been merged without proper styling description, see: https://github.com/wix/stylable-components/blob/master/CONTRIBUTING.md#bad-example
Please be careful next time ☮️

**Pull request merged**